File: exJava/go/Q24.go
package main
import "fmt"

func main() {
  ohm := newOrderedHashMap()
  ohm.put("AAA", 100)
  ohm.put("BBB", 200)
  ohm.put("CCC", 300)
  ohm.put("DDD", 400)
  ohm.put("EEE", 500)
  ohm.put("DDD", 1000)
  ohm.put("DDD", 2000)
  ohm.put("XYX", 10000)
  
  keyList := ohm.KeyList()
  for _, v := range keyList {
    fmt.Printf("%s = %d\n", v, ohm.get(v))
  }
}

type orderedHashMap struct {
  keys []string
  m map[string]int
}

func newOrderedHashMap() (ohm *orderedHashMap) {
  ohm = new(orderedHashMap)
  ohm.m = make(map[string]int)
  ohm.keys = make([]string, 0)
  return 
}
func (ohm *orderedHashMap) put(str string, value int) {
  if _, ok := ohm.m[str]; !ok {
    ohm.keys = append(ohm.keys, str)
  }
  ohm.m[str] = value
}
func (ohm *orderedHashMap) get(str string) int {
  return ohm.m[str]
}
func (ohm *orderedHashMap) KeyList() []string {
  return ohm.keys
}

File: exJava/java/Q24.java
import java.util.*;

public class Q24 {
  public static void main(String args[]) {
    new Q24Main();
  }
}

class Q24Main {

  public Q24Main() {
    OrderedHashMap omap = new OrderedHashMap();
    omap.put("AAA", 100);
    omap.put("BBB", 200);
    omap.put("CCC", 300);
    omap.put("DDD", 400);
    omap.put("EEE", 500);
    omap.put("DDD", 1000);
    omap.put("DDD", 2000);
    omap.put("XYX", 10000);

    List klist = omap.getKeyList();
    for (int i = 0; i < klist.size(); i++) {
      Object key = klist.get(i);
      System.out.println(key + " = " + omap.get(key));
    }
  }
}


class OrderedHashMap {

  public static final int L = 10;
  HashNode[] heads;
  HashNode ohead, otail; // ordered head and tail

  OrderedHashMap () {
    heads = new HashNode[L];
    ohead = null;
  }

  void put(Object key, Object value){
    HashNode node = new HashNode(key, value);
    int hash = node.key.hashCode() % L;
    hash = hash < 0 ? -hash : hash;
    HashNode head = heads[hash];
    while(head != null) {
      if (head.key.equals(node.key)) {
        head.value = node.value;
        break;
      } 
      head = head.next;
    }
    if (head == null) { // new node 
      node.next = heads[hash];
      heads[hash] = node;
      if (ohead == null) {
        ohead = otail = node;
      } else {
        otail.onext = node;
        otail = node;
      }
    }
  }

  Object get(Object key) {
    HashNode node = new HashNode(key, null);
    int hash = node.key.hashCode() % L;
    hash = hash < 0 ? -hash : hash;
    HashNode head = heads[hash];
    while(head != null) {
      if (head.key.equals(node.key)) {
        node.value = head.value;
        break;
      } 
      head = head.next;
    }
    return node.value;
  }

  List getKeyList() {
    HashNode hn = ohead;
    List<String> klist = new ArrayList<String>();
    while(hn != null) {
      klist.add(hn.key.toString());
      hn = hn.onext;
    }
    return klist;
  }
}

class HashNode {
  HashNode next, 
    onext; // ordered next
  Object key, value;

  public HashNode(Object key, Object value) {
    next = null;
    onext = null;
    this.key = key;
    this.value = value;
  }
}
